File: tobas_setup_assistant/src/tobas_setup_assistant/setting_widgets/magnetometer.py
from __future__ import annotations
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from ..setup_assistant import SetupAssistant
    from ..parameter_getters import ParamGetterWidget

from overrides import override
from PyQt5.QtWidgets import QVBoxLayout

from ..common import SENSOR_OFFSET_DESCRIPTION
from ..parameter_getters import ParamGetterWidget_SpinBox, ParamGetterWidget_Vector3d
from .base_setting import BaseSettingWidget


class MagnetometerWidget(BaseSettingWidget):
    NAME = "Magnetometer"
    TITLE_TEXT = "Define Magnetometer"
    ABST_TEXT = ""  # TODO

    def __init__(self, main: SetupAssistant) -> None:
        super().__init__(main)

        self._param_rows = QVBoxLayout()
        self._rows.addLayout(self._param_rows)

        self.offset = ParamGetterWidget_Vector3d("Offset", SENSOR_OFFSET_DESCRIPTION, suffix=" m")
        self._param_rows.addWidget(self.offset)

        update_rate_description = ""
        self.update_rate = ParamGetterWidget_SpinBox(
            "Update Rate", update_rate_description, minimum=1, default=80, suffix=" Hz"
        )
        self._param_rows.addWidget(self.update_rate)

        gauss_noise_description = ""
        self.gauss_noise = ParamGetterWidget_SpinBox(
            "Standard Deviation of Additive White Gaussian Noise",
            gauss_noise_description,
            minimum=0,
            default=80,
            suffix=" nT",
        )
        self._param_rows.addWidget(self.gauss_noise)

        uniform_noise_description = ""
        self.uniform_noise = ParamGetterWidget_SpinBox(
            "Symmetric Bounds of Uniform Noise for Initial Bias",
            uniform_noise_description,
            minimum=0,
            default=400,
            suffix=" nT",
        )
        self._param_rows.addWidget(self.uniform_noise)

        self._rows.addStretch()

    @override
    def update_internal_data_structures(self) -> None:
        pass

    @override
    def is_valid(self) -> bool:
        return True

    @override
    def dump_settings(self) -> dict:
        res = dict()
        for i in range(self._param_rows.count()):
            param: ParamGetterWidget = self._param_rows.itemAt(i).widget()
            res[param.name()] = param.get()
        return res

    @override
    def load_settings(self, data: dict) -> None:
        for i in range(self._param_rows.count()):
            param: ParamGetterWidget = self._param_rows.itemAt(i).widget()
            param.set(data[param.name()])

    def equipped(self) -> bool:
        return True

File: tobas_gui/tobas_setup_assistant/src/setting_widgets/magnetometer.cpp
#include "tobas_setup_assistant/setting_tabs/magnetometer.hpp"

#include <tobas_yaml_tools/convert/eigen.hpp>

namespace gui
{
namespace sa
{
MagnetometerWidget::MagnetometerWidget()
{
  offset_ = new ParamGetterWidget_Vector3d("Offset", kSensorOffsetDescription);
  offset_->setSuffix(" m");
  addWidget(offset_);

  update_rate_ = new ParamGetterWidget_SpinBox("Update Rate", "");  // TODO
  update_rate_->setMinimum(1);
  update_rate_->setValue(100);
  update_rate_->setSuffix(" Hz");
  addWidget(update_rate_);

  noise_stddev_ = new ParamGetterWidget_SpinBox("Standard Deviation of Additive White Gaussian Noise", "");  // TODO
  noise_stddev_->setMinimum(0);
  noise_stddev_->setValue(460);  // IIS2MDC
  noise_stddev_->setSuffix(" nT");
  addWidget(noise_stddev_);

  hard_bias_norm_ = new ParamGetterWidget_SpinBox("Hard-Iron Bias Norm", "");  // TODO
  hard_bias_norm_->setMinimum(0);
  hard_bias_norm_->setValue(3000);  // IIS2MDCの最大値は6000nTだが，キャリブレーションを前提としてそれより低めに設定．
  hard_bias_norm_->setSuffix(" nT");
  addWidget(hard_bias_norm_);

  addStretch();
}

const char* MagnetometerWidget::name() const
{
  return "Compass";
}

const char* MagnetometerWidget::title() const
{
  return "Define 3-Axis Magnetometer";
}

const char* MagnetometerWidget::description() const
{
  return "";  // TODO
}

void MagnetometerWidget::onOpened()
{
  return;
}

void MagnetometerWidget::updateInternalDataStructures()
{
  return;
}

bool MagnetometerWidget::isValid()
{
  return true;
}

YAML::Node MagnetometerWidget::dump() const
{
  YAML::Node node(YAML::NodeType::Map);

  node[offset_->name()] = offset_->getValue();
  node[update_rate_->name()] = update_rate_->getValue();
  node[noise_stddev_->name()] = noise_stddev_->getValue();
  node[hard_bias_norm_->name()] = hard_bias_norm_->getValue();

  return node;
}

void MagnetometerWidget::load(const YAML::Node& node)
{
  offset_->setValue(node[offset_->name()].as<Eigen::Vector3d>());
  update_rate_->setValue(node[update_rate_->name()].as<int>());
  noise_stddev_->setValue(node[noise_stddev_->name()].as<int>());
  hard_bias_norm_->setValue(node[hard_bias_norm_->name()].as<int>());
}

Eigen::Vector3d MagnetometerWidget::offset() const
{
  return offset_->getValue();
}

int MagnetometerWidget::updateRate() const
{
  return update_rate_->getValue();
}

int MagnetometerWidget::noiseStddev() const
{
  return noise_stddev_->getValue();
}

int MagnetometerWidget::hardBiasNorm() const
{
  return hard_bias_norm_->getValue();
}
}  // namespace sa
}  // namespace gui
